File: app/controllers/health_controller.rb
class HealthController < ActionController::API
  def show
    checks = {
      db: check_db,
      deepseek: check_deepseek,
      lihkg: check_lihkg
    }
    render_check(checks)
  end

  def db
    render_check(db: check_db)
  end

  def deepseek
    render_check(deepseek: check_deepseek)
  end

  def lihkg
    render_check(lihkg: check_lihkg)
  end

  private

  def render_check(checks)
    healthy = checks.values.all? { |check| check[:status] == "ok" }
    render json: { status: healthy ? "ok" : "error", checks: checks }, status: healthy ? :ok : :service_unavailable
  end

  def check_db
    result = ActiveRecord::Base.connection.select_value("PRAGMA journal_mode")
    wal = result.to_s.downcase == "wal"
    { status: wal ? "ok" : "error", wal: wal }
  rescue StandardError => e
    Rails.logger.warn("health db check failed: #{e.class}: #{e.message}")
    { status: "error", wal: false }
  end

  def check_deepseek
    response = upstream_connection(ENV.fetch("DEEPSEEK_BASE_URL", "https://api.deepseek.com"),
      "Bearer #{ENV.fetch("DEEPSEEK_API_KEY")}").get("/models")
    { status: response.success? ? "ok" : "error", code: response.status }
  rescue KeyError, Faraday::Error => e
    Rails.logger.warn("health deepseek check failed: #{e.class}: #{e.message}")
    { status: "error" }
  end

  def check_lihkg
    url = ENV.fetch("LIHKG_HEALTHCHECK_URL", ENV.fetch("LIHKG_UPLOAD_URL", "https://img.eservice-hk.net/api.php?version=2"))
    response = upstream_connection(url).get
    { status: response.success? ? "ok" : "error", code: response.status }
  rescue Faraday::Error => e
    Rails.logger.warn("health lihkg check failed: #{e.class}: #{e.message}")
    { status: "error" }
  end

  def upstream_connection(url, authorization = nil)
    Faraday.new(url:) do |faraday|
      faraday.options.timeout = 5
      faraday.options.open_timeout = 3
      faraday.headers["Authorization"] = authorization if authorization
      faraday.headers["Origin"] = "https://lihkg.com" if url.include?("img.eservice-hk.net")
    end
  end
end

File: src/health/health.controller.ts
import {Controller, Get, Res} from "@nestjs/common";
import type {Response} from "express";
import {DataSource} from "typeorm";

import {Public} from "../auth/public.decorator";
import {envOr} from "../config/env";

@Public()
@Controller()
export class HealthController {
    constructor(private readonly dataSource: DataSource) {}

    @Get("up")
    up(): void {
        // Rails 8 liveness: 200 with an empty body.
    }

    @Get("health")
    async health(@Res() response: Response): Promise<void> {
        this.render(
            response,
            await this.checks({
                db: true,
                deepseek: true,
                lihkg: true,
            })
        );
    }

    @Get("health/db")
    async dbHealth(@Res() response: Response): Promise<void> {
        this.render(response, {db: await this.checkDb()});
    }

    @Get("health/deepseek")
    async deepseekHealth(@Res() response: Response): Promise<void> {
        this.render(response, {deepseek: await this.checkDeepseek()});
    }

    @Get("health/lihkg")
    async lihkgHealth(@Res() response: Response): Promise<void> {
        this.render(response, {lihkg: await this.checkLihkg()});
    }

    private async checks(which: {db?: boolean; deepseek?: boolean; lihkg?: boolean}): Promise<Record<string, unknown>> {
        const checks: Record<string, unknown> = {};
        if (which.db) checks.db = await this.checkDb();
        if (which.deepseek) checks.deepseek = await this.checkDeepseek();
        if (which.lihkg) checks.lihkg = await this.checkLihkg();
        return checks;
    }

    private render(response: Response, checks: Record<string, unknown>): void {
        const healthy = Object.values(checks).every(check => (check as {status?: string}).status === "ok");
        response.status(healthy ? 200 : 503).json({
            status: healthy ? "ok" : "error",
            checks,
        });
    }

    private async checkDb(): Promise<Record<string, unknown>> {
        try {
            const rows = (await this.dataSource.query("PRAGMA journal_mode")) as Array<Record<string, unknown>>;
            const mode = typeof rows[0]?.journal_mode === "string" ? (rows[0].journal_mode as string) : "";
            const wal = mode.toLowerCase() === "wal";
            return {status: wal ? "ok" : "error", wal};
        } catch {
            return {status: "error", wal: false};
        }
    }

    private async checkDeepseek(): Promise<Record<string, unknown>> {
        const base = envOr("DEEPSEEK_BASE_URL", "https://api.deepseek.com");
        const key = process.env.DEEPSEEK_API_KEY;
        if (!key) {
            return {status: "error"};
        }
        try {
            const response = await fetch(`${base.replace(/\/+$/, "")}/models`, {
                headers: {Authorization: `Bearer ${key}`},
                signal: AbortSignal.timeout(5_000),
            });
            return {status: response.ok ? "ok" : "error", code: response.status};
        } catch {
            return {status: "error"};
        }
    }

    private async checkLihkg(): Promise<Record<string, unknown>> {
        const url = envOr("LIHKG_HEALTHCHECK_URL", "") || envOr("LIHKG_UPLOAD_URL", "https://img.eservice-hk.net/api.php?version=2");
        const headers: Record<string, string> = {};
        if (url.includes("img.eservice-hk.net")) {
            headers.Origin = "https://lihkg.com";
        }
        try {
            const response = await fetch(url, {headers, signal: AbortSignal.timeout(5_000)});
            return {status: response.ok ? "ok" : "error", code: response.status};
        } catch {
            return {status: "error"};
        }
    }
}
